File: lib/logistics.ts
/**
 * Deterministic logistics engine for Kazakhstan.
 *
 * The rates below are configurable market benchmarks, not carrier quotations.
 * Every quote therefore carries `rateKind: "benchmark"`; a future carrier API can
 * provide the same LogisticsQuote contract with `rateKind: "live"`.
 */

export interface City {
  id: string;
  kz: string;
  ru: string;
  lat: number;
  lon: number;
}

export const CITIES: City[] = [
  { id: "astana", kz: "Астана", ru: "Астана", lat: 51.169, lon: 71.449 },
  { id: "almaty", kz: "Алматы", ru: "Алматы", lat: 43.238, lon: 76.946 },
  { id: "shymkent", kz: "Шымкент", ru: "Шымкент", lat: 42.315, lon: 69.587 },
  { id: "karaganda", kz: "Қарағанды", ru: "Караганда", lat: 49.806, lon: 73.085 },
  { id: "aktobe", kz: "Ақтөбе", ru: "Актобе", lat: 50.283, lon: 57.167 },
  { id: "taraz", kz: "Тараз", ru: "Тараз", lat: 42.9, lon: 71.367 },
  { id: "pavlodar", kz: "Павлодар", ru: "Павлодар", lat: 52.287, lon: 76.967 },
  { id: "oskemen", kz: "Өскемен", ru: "Усть-Каменогорск", lat: 49.948, lon: 82.628 },
  { id: "semey", kz: "Семей", ru: "Семей", lat: 50.411, lon: 80.227 },
  { id: "atyrau", kz: "Атырау", ru: "Атырау", lat: 47.107, lon: 51.903 },
  { id: "kostanay", kz: "Қостанай", ru: "Костанай", lat: 53.214, lon: 63.624 },
  { id: "kyzylorda", kz: "Қызылорда", ru: "Кызылорда", lat: 44.853, lon: 65.509 },
  { id: "oral", kz: "Орал", ru: "Уральск", lat: 51.233, lon: 51.367 },
  { id: "petropavl", kz: "Петропавл", ru: "Петропавловск", lat: 54.865, lon: 69.136 },
  { id: "aktau", kz: "Ақтау", ru: "Актау", lat: 43.65, lon: 51.16 },
  { id: "taldykorgan", kz: "Талдықорған", ru: "Талдыкорган", lat: 45.017, lon: 78.373 },
  { id: "turkistan", kz: "Түркістан", ru: "Туркестан", lat: 43.297, lon: 68.251 },
  { id: "kokshetau", kz: "Көкшетау", ru: "Кокшетау", lat: 53.283, lon: 69.383 },
];

export const cityById = (id: string) => CITIES.find((c) => c.id === id);

export interface DistanceProvider {
  name: string;
  /** Road distance in km. */
  distanceKm(fromId: string, toId: string): number;
}

export type LogisticsMode = "city" | "truck" | "gazelle" | "rail" | "air";
export type TransportMode = "auto" | LogisticsMode;

export interface LogisticsQuote {
  mode: LogisticsMode;
  distanceKm: number;
  units: number;
  capacityTonnes: number | null;
  cost: number;
  transitDays: number;
  rateKind: "benchmark" | "live";
  rateLabel: string;
}

export interface LogisticsPlan {
  selected: LogisticsQuote;
  recommendedMode: LogisticsMode;
  quotes: LogisticsQuote[];
  /** Normal road transit used to measure faster/slower explicit scenarios. */
  roadReferenceDays: number;
}

/**
 * Carrier-market rate for one road mode and route, fetched server-side from ATI.SU
 * (average rate over the last 30 days). The engine only uses it for the same route.
 */
export interface LiveRoadRate {
  mode: "truck" | "gazelle";
  fromId: string;
  toId: string;
  distanceKm: number;
  /** Market price of one trip of one vehicle, KZT */
  perTripKzt: number;
  source: "ati.su";
  fetchedAt: string;
}

export const LIVE_RATE_LABEL = "Verified Live (ATI.SU)";
export const ESTIMATED_RATE_LABEL = "Estimated Rate";

export const LOGISTICS_RATES = {
  roadFactor: 1.25,
  railFactor: 1.1,
  truck: { capacityT: 20, perKmKzt: 450, returnShare: 0.5, fuelShare: 0.4, minimumKzt: 60_000, kmPerDay: 650 },
  gazelle: { capacityT: 3, perKmKzt: 180, fuelShare: 0.4, minimumKzt: 30_000, kmPerDay: 550 },
  rail: { capacityT: 20, perTonnePer1000KmKzt: 12_000, terminalPerContainerKzt: 120_000, minimumBillableT: 10, kmPerDay: 450, terminalDays: 2 },
  air: { perKgKzt: 850, minimumKzt: 60_000, transitDays: 1 },
  city: { capacityT: 3, perTripKzt: 25_000, heavyTripKzt: 60_000 },
} as const;

const cityPair = (fromId: string, toId: string) => {
  const a = cityById(fromId);
  const b = cityById(toId);
  if (!a || !b) throw new Error(`Unknown city: ${!a ? fromId : toId}`);
  return { a, b };
};

/** Straight-line great-circle distance, before transport-specific coefficients. */
export function directDistanceKm(fromId: string, toId: string): number {
  if (fromId === toId) return 0;
  const { a, b } = cityPair(fromId, toId);
  const radius = 6371;
  const toRad = (degrees: number) => (degrees * Math.PI) / 180;
  const dLat = toRad(b.lat - a.lat);
  const dLon = toRad(b.lon - a.lon);
  const h = Math.sin(dLat / 2) ** 2 + Math.cos(toRad(a.lat)) * Math.cos(toRad(b.lat)) * Math.sin(dLon / 2) ** 2;
  return 2 * radius * Math.asin(Math.sqrt(h));
}

export const haversineProvider: DistanceProvider = {
  name: "haversine×1.25",
  distanceKm(fromId, toId) {
    if (fromId === toId) return 15;
    return Math.round(directDistanceKm(fromId, toId) * LOGISTICS_RATES.roadFactor);
  },
};

/** Backwards-compatible standard 20-t truck constants. */
export const FREIGHT = {
  truckCapacityT: LOGISTICS_RATES.truck.capacityT,
  tariffPerKm: LOGISTICS_RATES.truck.perKmKzt,
  returnShare: LOGISTICS_RATES.truck.returnShare,
  fuelShare: LOGISTICS_RATES.truck.fuelShare,
  minPerTrip: LOGISTICS_RATES.truck.minimumKzt,
} as const;

export interface FreightQuote {
  distanceKm: number;
  trucks: number;
  cost: number;
}

/** Backwards-compatible explicit truck calculation. */
export function freightCost(distanceKm: number, tonnes: number, fuelDeltaPct = 0, transportDeltaPct = 0): FreightQuote {
  const trucks = Math.max(1, Math.ceil(Math.max(0, tonnes) / LOGISTICS_RATES.truck.capacityT));
  const tariff =
    LOGISTICS_RATES.truck.perKmKzt *
    (1 + LOGISTICS_RATES.truck.fuelShare * (fuelDeltaPct / 100)) *
    (1 + transportDeltaPct / 100);
  const perTrip = Math.max(LOGISTICS_RATES.truck.minimumKzt, distanceKm * tariff * (1 + LOGISTICS_RATES.truck.returnShare));
  return { distanceKm, trucks, cost: trucks * perTrip };
}

function quoteForMode(fromId: string, toId: string, tonnes: number, mode: LogisticsMode, fuelDeltaPct: number, transportDeltaPct: number, live?: LiveRoadRate): LogisticsQuote {
  const cargo = Math.max(0.001, tonnes);
  const sameCity = fromId === toId;
  const direct = directDistanceKm(fromId, toId);
  const roadKm = sameCity ? 15 : Math.round(direct * LOGISTICS_RATES.roadFactor);
  const railKm = sameCity ? 15 : Math.round(direct * LOGISTICS_RATES.railFactor);
  const airKm = sameCity ? 15 : Math.round(direct);
  const market = Math.max(0.1, 1 + transportDeltaPct / 100);

  if (mode === "city") {
    const capacity = LOGISTICS_RATES.city.capacityT;
    const units = Math.max(1, Math.ceil(cargo / capacity));
    const perTrip = cargo <= capacity ? LOGISTICS_RATES.city.perTripKzt : LOGISTICS_RATES.city.heavyTripKzt;
    return { mode, distanceKm: 15, units, capacityTonnes: capacity, cost: units * perTrip * market, transitDays: 1, rateKind: "benchmark", rateLabel: "fixed city rate" };
  }
  // Live ATI.SU market rate for this exact route replaces the benchmark tariff.
  if ((mode === "truck" || mode === "gazelle") && live && live.mode === mode && live.fromId === fromId && live.toId === toId) {
    const spec = mode === "truck" ? LOGISTICS_RATES.truck : LOGISTICS_RATES.gazelle;
    const units = Math.max(1, Math.ceil(cargo / spec.capacityT));
    const adjust = (1 + spec.fuelShare * (fuelDeltaPct / 100)) * market;
    return {
      mode,
      distanceKm: live.distanceKm,
      units,
      capacityTonnes: spec.capacityT,
      cost: units * live.perTripKzt * adjust,
      transitDays: Math.max(1, Math.ceil(live.distanceKm / spec.kmPerDay)),
      rateKind: "live",
      rateLabel: LIVE_RATE_LABEL,
    };
  }
  if (mode === "truck") {
    const q = freightCost(roadKm, cargo, fuelDeltaPct, transportDeltaPct);
    return { mode, distanceKm: roadKm, units: q.trucks, capacityTonnes: LOGISTICS_RATES.truck.capacityT, cost: q.cost, transitDays: Math.max(1, Math.ceil(roadKm / LOGISTICS_RATES.truck.kmPerDay)), rateKind: "benchmark", rateLabel: "450 KZT/km + 50% return" };
  }
  if (mode === "gazelle") {
    const units = Math.max(1, Math.ceil(cargo / LOGISTICS_RATES.gazelle.capacityT));
    const tariff = LOGISTICS_RATES.gazelle.perKmKzt * (1 + LOGISTICS_RATES.gazelle.fuelShare * fuelDeltaPct / 100) * market;
    const cost = units * Math.max(LOGISTICS_RATES.gazelle.minimumKzt, roadKm * tariff);
    return { mode, distanceKm: roadKm, units, capacityTonnes: LOGISTICS_RATES.gazelle.capacityT, cost, transitDays: Math.max(1, Math.ceil(roadKm / LOGISTICS_RATES.gazelle.kmPerDay)), rateKind: "benchmark", rateLabel: "180 KZT/km" };
  }
  if (mode === "rail") {
    const units = Math.max(1, Math.ceil(cargo / LOGISTICS_RATES.rail.capacityT));
    const billableTonnes = Math.max(LOGISTICS_RATES.rail.minimumBillableT, cargo);
    const lineHaul = billableTonnes * LOGISTICS_RATES.rail.perTonnePer1000KmKzt * (railKm / 1000);
    const terminals = units * LOGISTICS_RATES.rail.terminalPerContainerKzt;
    return { mode, distanceKm: railKm, units, capacityTonnes: LOGISTICS_RATES.rail.capacityT, cost: (lineHaul + terminals) * market, transitDays: Math.max(3, Math.ceil(railKm / LOGISTICS_RATES.rail.kmPerDay) + LOGISTICS_RATES.rail.terminalDays), rateKind: "benchmark", rateLabel: "12,000 KZT/t per 1,000 km + terminals" };
  }
  return { mode: "air", distanceKm: airKm, units: 1, capacityTonnes: null, cost: Math.max(LOGISTICS_RATES.air.minimumKzt, cargo * 1000 * LOGISTICS_RATES.air.perKgKzt) * market, transitDays: LOGISTICS_RATES.air.transitDays, rateKind: "benchmark", rateLabel: "850 KZT/kg" };
}

/** All usable modes for this route. Same-city routes intentionally collapse to city delivery. */
export function logisticsQuotes(fromId: string, toId: string, tonnes: number, fuelDeltaPct = 0, transportDeltaPct = 0, liveRates: LiveRoadRate[] = []): LogisticsQuote[] {
  cityPair(fromId, toId);
  if (fromId === toId) return [quoteForMode(fromId, toId, tonnes, "city", fuelDeltaPct, transportDeltaPct)];
  return (["truck", "gazelle", "rail", "air"] as LogisticsMode[]).map((mode) =>
    quoteForMode(fromId, toId, tonnes, mode, fuelDeltaPct, transportDeltaPct, liveRates.find((r) => r.mode === mode && r.fromId === fromId && r.toId === toId))
  );
}

export function recommendLogisticsMode(quotes: LogisticsQuote[], tonnes: number, deliveryDays: number): LogisticsMode {
  if (quotes.length === 1) return quotes[0].mode;
  const by = (mode: LogisticsMode) => quotes.find((q) => q.mode === mode)!;
  if (deliveryDays <= 2) return "air";
  if (tonnes <= 3) return "gazelle";
  if (tonnes >= 10 && by("rail").cost < by("truck").cost && by("rail").transitDays <= Math.max(3, deliveryDays)) return "rail";
  return "truck";
}

export function logisticsPlan(fromId: string, toId: string, tonnes: number, requested: TransportMode = "auto", deliveryDays = 30, fuelDeltaPct = 0, transportDeltaPct = 0, liveRates: LiveRoadRate[] = []): LogisticsPlan {
  const quotes = logisticsQuotes(fromId, toId, tonnes, fuelDeltaPct, transportDeltaPct, liveRates);
  const recommendedMode = recommendLogisticsMode(quotes, tonnes, deliveryDays);
  const allowed = quotes.some((q) => q.mode === requested) ? requested as LogisticsMode : recommendedMode;
  const selected = quotes.find((q) => q.mode === allowed) ?? quotes[0];
  const roadReference = quotes.find((q) => q.mode === (tonnes <= 3 ? "gazelle" : "truck")) ?? quotes[0];
  return { selected, recommendedMode, quotes, roadReferenceDays: roadReference.transitDays };
}

File: bot/logistics.py
"""Deterministic Kazakhstan logistics quotes shared by the Telegram economic layer.

Built-in rates are explicit market benchmarks ("Estimated Rate"). For road modes
(20-t truck, Gazelle) a live ATI.SU average market rate can be fetched with
fetch_ati_live_rates() and passed in; the quote then becomes "Verified Live (ATI.SU)".
The calculation itself stays deterministic: the same inputs, including the live rate,
always give the same result.
"""
from __future__ import annotations

import datetime as _dt
import math
import os
import time
from dataclasses import dataclass
from typing import Dict, List, Literal, Optional, Sequence, Tuple

import httpx
from pydantic import BaseModel

LogisticsMode = Literal["city", "truck", "gazelle", "rail", "air"]
TransportMode = Literal["auto", "city", "truck", "gazelle", "rail", "air"]


@dataclass(frozen=True)
class City:
    id: str
    kz: str
    ru: str
    lat: float
    lon: float


CITIES: List[City] = [
    City("astana", "Астана", "Астана", 51.169, 71.449),
    City("almaty", "Алматы", "Алматы", 43.238, 76.946),
    City("shymkent", "Шымкент", "Шымкент", 42.315, 69.587),
    City("karaganda", "Қарағанды", "Караганда", 49.806, 73.085),
    City("aktobe", "Ақтөбе", "Актобе", 50.283, 57.167),
    City("taraz", "Тараз", "Тараз", 42.9, 71.367),
    City("pavlodar", "Павлодар", "Павлодар", 52.287, 76.967),
    City("oskemen", "Өскемен", "Усть-Каменогорск", 49.948, 82.628),
    City("semey", "Семей", "Семей", 50.411, 80.227),
    City("atyrau", "Атырау", "Атырау", 47.107, 51.903),
    City("kostanay", "Қостанай", "Костанай", 53.214, 63.624),
    City("kyzylorda", "Қызылорда", "Кызылорда", 44.853, 65.509),
    City("oral", "Орал", "Уральск", 51.233, 51.367),
    City("petropavl", "Петропавл", "Петропавловск", 54.865, 69.136),
    City("aktau", "Ақтау", "Актау", 43.65, 51.16),
    City("taldykorgan", "Талдықорған", "Талдыкорган", 45.017, 78.373),
    City("turkistan", "Түркістан", "Туркестан", 43.297, 68.251),
    City("kokshetau", "Көкшетау", "Кокшетау", 53.283, 69.383),
]
CITY_BY_ID: Dict[str, City] = {city.id: city for city in CITIES}


class FREIGHT:
    ROAD_FACTOR = 1.25
    RAIL_FACTOR = 1.10
    TRUCK_CAPACITY_T = 20.0
    TARIFF_PER_KM = 450.0
    RETURN_SHARE = 0.5
    FUEL_SHARE = 0.4
    MIN_PER_TRIP = 60_000.0
    TRUCK_KM_PER_DAY = 650.0
    GAZELLE_CAPACITY_T = 3.0
    GAZELLE_PER_KM = 180.0
    GAZELLE_MIN = 30_000.0
    GAZELLE_KM_PER_DAY = 550.0
    RAIL_CAPACITY_T = 20.0
    RAIL_PER_T_PER_1000_KM = 12_000.0
    RAIL_TERMINAL_PER_CONTAINER = 120_000.0
    RAIL_MIN_BILLABLE_T = 10.0
    RAIL_KM_PER_DAY = 450.0
    AIR_PER_KG = 850.0
    AIR_MIN = 60_000.0
    CITY_CAPACITY_T = 3.0
    CITY_TRIP = 25_000.0
    CITY_HEAVY_TRIP = 60_000.0


LIVE_RATE_LABEL = "Verified Live (ATI.SU)"
ESTIMATED_RATE_LABEL = "Estimated Rate"


class LiveRoadRate(BaseModel):
    """Средняя рыночная ставка ATI.SU за рейс одной машины на конкретном маршруте, ₸."""

    mode: Literal["truck", "gazelle"]
    from_id: str
    to_id: str
    distance_km: int
    per_trip_kzt: float
    source: Literal["ati.su"] = "ati.su"
    fetched_at: str = ""


@dataclass(frozen=True)
class FreightQuote:
    distance_km: float
    trucks: int
    cost: float


@dataclass(frozen=True)
class LogisticsQuote:
    mode: LogisticsMode
    distance_km: int
    units: int
    capacity_tonnes: Optional[float]
    cost: float
    transit_days: int
    rate_kind: Literal["benchmark", "live"] = "benchmark"
    rate_label: str = ""


@dataclass(frozen=True)
class LogisticsPlan:
    selected: LogisticsQuote
    recommended_mode: LogisticsMode
    quotes: List[LogisticsQuote]
    road_reference_days: int


def _js_round(value: float) -> int:
    return int(math.floor(value + 0.5))


def direct_distance_km(from_id: str, to_id: str) -> float:
    if from_id == to_id:
        return 0.0
    a, b = CITY_BY_ID.get(from_id), CITY_BY_ID.get(to_id)
    if a is None or b is None:
        raise ValueError(f"Unknown city: {from_id if a is None else to_id}")
    radius = 6371.0
    rad = lambda degrees: degrees * math.pi / 180  # noqa: E731
    d_lat, d_lon = rad(b.lat - a.lat), rad(b.lon - a.lon)
    h = math.sin(d_lat / 2) ** 2 + math.cos(rad(a.lat)) * math.cos(rad(b.lat)) * math.sin(d_lon / 2) ** 2
    return 2 * radius * math.asin(math.sqrt(h))


def distance_km(from_id: str, to_id: str) -> int:
    """Road distance: Haversine × 1.25; local delivery uses a 15 km allowance."""
    return 15 if from_id == to_id else _js_round(direct_distance_km(from_id, to_id) * FREIGHT.ROAD_FACTOR)


def freight_cost(dist_km: float, tonnes: float, fuel_delta_pct: float = 0.0, transport_delta_pct: float = 0.0) -> FreightQuote:
    trucks = max(1, math.ceil(max(0.0, tonnes) / FREIGHT.TRUCK_CAPACITY_T))
    tariff = FREIGHT.TARIFF_PER_KM * (1 + FREIGHT.FUEL_SHARE * fuel_delta_pct / 100) * (1 + transport_delta_pct / 100)
    per_trip = max(FREIGHT.MIN_PER_TRIP, dist_km * tariff * (1 + FREIGHT.RETURN_SHARE))
    return FreightQuote(distance_km=dist_km, trucks=trucks, cost=trucks * per_trip)


def _quote(
    from_id: str,
    to_id: str,
    tonnes: float,
    mode: LogisticsMode,
    fuel_delta_pct: float,
    transport_delta_pct: float,
    live: Optional[LiveRoadRate] = None,
) -> LogisticsQuote:
    cargo = max(0.001, tonnes)
    same_city = from_id == to_id
    direct = direct_distance_km(from_id, to_id)
    road_km = 15 if same_city else _js_round(direct * FREIGHT.ROAD_FACTOR)
    rail_km = 15 if same_city else _js_round(direct * FREIGHT.RAIL_FACTOR)
    air_km = 15 if same_city else _js_round(direct)
    market = max(0.1, 1 + transport_delta_pct / 100)

    if mode == "city":
        units = max(1, math.ceil(cargo / FREIGHT.CITY_CAPACITY_T))
        per_trip = FREIGHT.CITY_TRIP if cargo <= FREIGHT.CITY_CAPACITY_T else FREIGHT.CITY_HEAVY_TRIP
        return LogisticsQuote(mode, 15, units, FREIGHT.CITY_CAPACITY_T, units * per_trip * market, 1, rate_label="fixed city rate")
    # Живая ставка ATI.SU для этого маршрута заменяет эталонный тариф
    if mode in ("truck", "gazelle") and live is not None and live.mode == mode and live.from_id == from_id and live.to_id == to_id:
        capacity = FREIGHT.TRUCK_CAPACITY_T if mode == "truck" else FREIGHT.GAZELLE_CAPACITY_T
        km_per_day = FREIGHT.TRUCK_KM_PER_DAY if mode == "truck" else FREIGHT.GAZELLE_KM_PER_DAY
        units = max(1, math.ceil(cargo / capacity))
        adjust = (1 + FREIGHT.FUEL_SHARE * fuel_delta_pct / 100) * market
        return LogisticsQuote(
            mode, live.distance_km, units, capacity, units * live.per_trip_kzt * adjust,
            max(1, math.ceil(live.distance_km / km_per_day)), rate_kind="live", rate_label=LIVE_RATE_LABEL,
        )
    if mode == "truck":
        q = freight_cost(road_km, cargo, fuel_delta_pct, transport_delta_pct)
        return LogisticsQuote(mode, road_km, q.trucks, FREIGHT.TRUCK_CAPACITY_T, q.cost, max(1, math.ceil(road_km / FREIGHT.TRUCK_KM_PER_DAY)), rate_label="450 KZT/km + 50% return")
    if mode == "gazelle":
        units = max(1, math.ceil(cargo / FREIGHT.GAZELLE_CAPACITY_T))
        tariff = FREIGHT.GAZELLE_PER_KM * (1 + FREIGHT.FUEL_SHARE * fuel_delta_pct / 100) * market
        cost = units * max(FREIGHT.GAZELLE_MIN, road_km * tariff)
        return LogisticsQuote(mode, road_km, units, FREIGHT.GAZELLE_CAPACITY_T, cost, max(1, math.ceil(road_km / FREIGHT.GAZELLE_KM_PER_DAY)), rate_label="180 KZT/km")
    if mode == "rail":
        units = max(1, math.ceil(cargo / FREIGHT.RAIL_CAPACITY_T))
        billable = max(FREIGHT.RAIL_MIN_BILLABLE_T, cargo)
        cost = (billable * FREIGHT.RAIL_PER_T_PER_1000_KM * rail_km / 1000 + units * FREIGHT.RAIL_TERMINAL_PER_CONTAINER) * market
        return LogisticsQuote(mode, rail_km, units, FREIGHT.RAIL_CAPACITY_T, cost, max(3, math.ceil(rail_km / FREIGHT.RAIL_KM_PER_DAY) + 2), rate_label="12,000 KZT/t per 1,000 km + terminals")
    cost = max(FREIGHT.AIR_MIN, cargo * 1000 * FREIGHT.AIR_PER_KG) * market
    return LogisticsQuote("air", air_km, 1, None, cost, 1, rate_label="850 KZT/kg")


def logistics_quotes(
    from_id: str,
    to_id: str,
    tonnes: float,
    fuel_delta_pct: float = 0.0,
    transport_delta_pct: float = 0.0,
    live_rates: Sequence[LiveRoadRate] = (),
) -> List[LogisticsQuote]:
    if from_id not in CITY_BY_ID or to_id not in CITY_BY_ID:
        raise ValueError(f"Unknown city: {from_id if from_id not in CITY_BY_ID else to_id}")
    if from_id == to_id:
        return [_quote(from_id, to_id, tonnes, "city", fuel_delta_pct, transport_delta_pct)]
    def live_for(mode: str) -> Optional[LiveRoadRate]:
        return next((r for r in live_rates if r.mode == mode and r.from_id == from_id and r.to_id == to_id), None)

    return [_quote(from_id, to_id, tonnes, mode, fuel_delta_pct, transport_delta_pct, live_for(mode)) for mode in ("truck", "gazelle", "rail", "air")]


def recommend_logistics_mode(quotes: List[LogisticsQuote], tonnes: float, delivery_days: int) -> LogisticsMode:
    if len(quotes) == 1:
        return quotes[0].mode
    by = {quote.mode: quote for quote in quotes}
    if delivery_days <= 2:
        return "air"
    if tonnes <= 3:
        return "gazelle"
    if tonnes >= 10 and by["rail"].cost < by["truck"].cost and by["rail"].transit_days <= max(3, delivery_days):
        return "rail"
    return "truck"


def logistics_plan(
    from_id: str,
    to_id: str,
    tonnes: float,
    requested: TransportMode = "auto",
    delivery_days: int = 30,
    fuel_delta_pct: float = 0.0,
    transport_delta_pct: float = 0.0,
    live_rates: Sequence[LiveRoadRate] = (),
) -> LogisticsPlan:
    quotes = logistics_quotes(from_id, to_id, tonnes, fuel_delta_pct, transport_delta_pct, live_rates)
    recommended = recommend_logistics_mode(quotes, tonnes, delivery_days)
    selected = next((quote for quote in quotes if quote.mode == requested), next(quote for quote in quotes if quote.mode == recommended))
    road_mode: LogisticsMode = "gazelle" if tonnes <= 3 else "truck"
    road = next((quote for quote in quotes if quote.mode == road_mode), quotes[0])
    return LogisticsPlan(selected=selected, recommended_mode=recommended, quotes=quotes, road_reference_days=road.transit_days)


# =========================================================================== #
# ATI.SU — живые рыночные ставки автоперевозок (Фура 20 т, Газель 3 т)          #
# =========================================================================== #

ATI_AVERAGE_PRICES_URL = "https://api.ati.su/priceline/license/v1/average_prices"

# ID городов в справочнике ATI.SU (получены через /gw/gis-dict/v1/cities/by-coordinate).
ATI_CITY_IDS: Dict[str, int] = {
    "astana": 3100, "almaty": 1004, "shymkent": 3148, "karaganda": 3103, "aktobe": 1005, "taraz": 3147,
    "pavlodar": 3055, "oskemen": 3058, "semey": 12209, "atyrau": 3048, "kostanay": 131, "kyzylorda": 3093,
    "oral": 1116, "petropavl": 2558, "aktau": 512, "taldykorgan": 23806, "turkistan": 3149, "kokshetau": 3101,
}

# Тентованный кузов: фура 20 т и газель 3 т (допустимые тоннажи API: 1.5 | 3 | 5 | 10 | 20)
ATI_AUTO_PROFILES: Dict[str, Dict[str, object]] = {
    "truck": {"CarType": "tent", "Tonnage": 20},
    "gazelle": {"CarType": "tent", "Tonnage": 3},
}

# API отдаёт цены в рублях (PricesInRub). Курс ₽→₸ — настройка, а не рыночные данные:
# задайте ATI_RUB_TO_KZT актуальным курсом.
DEFAULT_RUB_TO_KZT = 6.3

AtiStatus = Literal["live", "no-key", "no-license", "rate-limit", "invalid-key", "http-error", "network", "no-data", "unknown-city"]


@dataclass(frozen=True)
class AtiRateResult:
    status: AtiStatus
    rate: Optional[LiveRoadRate] = None
    bottom_kzt: float = 0.0
    upper_kzt: float = 0.0
    loads_count: int = 0
    detail: str = ""


_ATI_CACHE: Dict[Tuple[str, str, str], Tuple[float, AtiRateResult]] = {}
_ATI_TTL_LIVE = 6 * 3600
_ATI_TTL_FAIL = 600  # после оплаты лицензии живые ставки подхватятся в течение 10 минут


def clear_ati_cache() -> None:
    _ATI_CACHE.clear()


def _rub_to_kzt() -> float:
    try:
        value = float(os.getenv("ATI_RUB_TO_KZT") or DEFAULT_RUB_TO_KZT)
    except ValueError:
        return DEFAULT_RUB_TO_KZT
    return value if value > 0 else DEFAULT_RUB_TO_KZT


def _parse_ati(response: httpx.Response, city_from: str, city_to: str, truck_type: str) -> AtiRateResult:
    code = response.status_code
    if code == 401:
        return AtiRateResult("invalid-key")
    if code == 403:
        return AtiRateResult("no-license", detail="ATI.SU: лицензия «Средние ставки» не активна")
    if code == 429:
        return AtiRateResult("rate-limit")
    if code >= 400:
        return AtiRateResult("http-error", detail=f"HTTP {code}")
    try:
        payload = response.json()
    except ValueError:
        return AtiRateResult("http-error", detail="bad json")
    rows = [row for row in (payload.get("Data") or []) if ((row.get("PricesInRub") or {}).get("AveragePrice") or 0) > 0]
    if not rows:
        return AtiRateResult("no-data")
    last = rows[-1]
    prices = last["PricesInRub"]
    distance = int(round(payload.get("Distance") or 0)) or distance_km(city_from, city_to)
    rate = _rub_to_kzt()

    def per_trip(value: float) -> float:
        # Небольшое значение — ставка за км, крупное — за рейс
        return (value * distance if value < 1000 else value) * rate

    live = LiveRoadRate(
        mode=truck_type,  # type: ignore[arg-type]
        from_id=city_from,
        to_id=city_to,
        distance_km=distance,
        per_trip_kzt=round(per_trip(float(prices["AveragePrice"]))),
        fetched_at=_dt.datetime.now(_dt.timezone.utc).isoformat(timespec="seconds"),
    )
    return AtiRateResult(
        "live",
        rate=live,
        bottom_kzt=round(per_trip(float(prices.get("BottomPrice") or 0))),
        upper_kzt=round(per_trip(float(prices.get("UpperPrice") or 0))),
        loads_count=int(last.get("LoadsCount") or 0),
    )


async def fetch_ati_live_rates(
    city_from: str,
    city_to: str,
    truck_type: str,
    api_key: Optional[str] = None,
    transport: Optional[httpx.AsyncBaseTransport] = None,
    timeout: float = 8.0,
) -> AtiRateResult:
    """Средняя рыночная ставка ATI.SU за последние 30 дней (Фура 20 т или Газель 3 т).

    Никогда не бросает исключений: при отсутствии ключа, лицензии, лимите или сетевой
    ошибке возвращает статус, и расчёт мягко остаётся на встроенном тарифе (Estimated Rate).
    """
    if truck_type not in ATI_AUTO_PROFILES:
        raise ValueError(f"truck_type must be truck or gazelle, got {truck_type!r}")
    key = (api_key if api_key is not None else os.getenv("ATI_SU_API_KEY", "")).strip()
    if not key:
        return AtiRateResult("no-key")
    a, b = ATI_CITY_IDS.get(city_from), ATI_CITY_IDS.get(city_to)
    if a is None or b is None or city_from == city_to:
        return AtiRateResult("unknown-city")

    cache_key = (city_from, city_to, truck_type)
    cached = _ATI_CACHE.get(cache_key)
    if cached and cached[0] > time.monotonic():
        return cached[1]

    today = _dt.date.today()
    body = {
        "From": {"CityId": a},
        "To": {"CityId": b},
        **ATI_AUTO_PROFILES[truck_type],
        "DateFrom": (today - _dt.timedelta(days=30)).isoformat(),
        "DateTo": today.isoformat(),
        "Frequency": "month",
        "WithNds": False,
        "RoundTrip": False,
    }
    try:
        async with httpx.AsyncClient(timeout=timeout, transport=transport) as client:
            response = await client.post(ATI_AVERAGE_PRICES_URL, json=body, headers={"Authorization": f"Bearer {key}"})
    except httpx.HTTPError as e:
        return AtiRateResult("network", detail=type(e).__name__)  # сеть не кэшируем
    result = _parse_ati(response, city_from, city_to, truck_type)
    _ATI_CACHE[cache_key] = (time.monotonic() + (_ATI_TTL_LIVE if result.status == "live" else _ATI_TTL_FAIL), result)
    return result


async def fetch_route_live_rates(city_from: str, city_to: str, api_key: Optional[str] = None) -> List[LiveRoadRate]:
    """Живые ставки фуры и газели для маршрута; пустой список → Estimated Rate."""
    if city_from == city_to:
        return []
    results = [await fetch_ati_live_rates(city_from, city_to, mode, api_key=api_key) for mode in ("truck", "gazelle")]
    return [r.rate for r in results if r.status == "live" and r.rate is not None]
